### DISCUSSION TITLE
ImGui special_key assertion failure.

### DISCUSSION THREAD

**Discussion by kannode** [vsg-dev/VulkanSceneGraph#630] [General]
Typing into any input box in ImGui consistently results in the below assertion failure on Linux and mac and likely on widows as well.

Assertion failed: (special_key > 256 && "ASCII stop at 127, but we use the range [257, 511]"), function _mapToSpecialKey, file SendEventsToImGui.cpp, line 107.

Am I missing something?
I searched issues and discussions but it seems like no one else is having this issue.

**Comment by robertosfield** [comment]
What are you typing into the input box?

Does the problem happen with the vsgimgui example?

I have just tried the demo window in the vsgimgui example enabling the text input and have tried all the combinations and nothing crashes.  Some of the behaviour seems a bit laggy, but I presume this is an ImGui issue.

**Reply by kannode** [reply]
Yes, the issue occurs with vsgimgui_example as well.  To replicate, click the "Demo Window".

- In the demo window go to Widgets/Basic and go to the "Hello World" input string and type anything in and you will get an assertion failure. This happens with any widget that has an input. It seems the "special_key" is set to 0 for some reason. See attached snap shot.
- There is no lag on Windows, or Mac, but does lag on Xwindows.
- Same issue across all platforms for vsgimgui.

**Reply by kannode** [reply]
[
![Capture](https://user-images.githubusercontent.com/32265121/205760523-ffd95af7-934e-4542-b1a2-4b5d3fc5a4ec.PNG)
![Capture2](https://user-images.githubusercontent.com/32265121/205760533-76f1b098-0154-4c8b-8eb5-790057a9726c.PNG)
](url)

**Comment by rolandhill** [comment]
Hi Kannode,

Are you typing on the number pad by any chance? I've got a fix - I need to
clean it up and submit it as a PR.

Regards,
Roland

**Reply by kannode** [reply]
Roland, No, I am not typing on the keypad.

**Comment by rolandhill** [comment]
Hi Robert and Kannode,

I just sent PR#30 that should fix this.
https://github.com/vsg-dev/vsgImGui/pull/30

Regards,
Roland

**Reply by kannode** [reply]
Hi @robertosfield @rolandhill 

I may have found the issue, but do not know where it originates.
![Capture4](https://user-images.githubusercontent.com/32265121/205762259-87869596-389c-4062-96c0-70eefe96aff7.PNG)

it seems after I pressed they key "a", _vsgToIntermediateMap cannot find the keyAndModifier in line 103, itr is thus "end" as seen at the bottom of the snapshot, resulting in special_key becoming 0 due to line 104... and subsequently the assertion fails. Happy to continue looking at this but I feel if someone else has a better understanding of the mapToSpecialKey function, they would have a better shot a fixing it.

**Comment by rolandhill** [comment]
I have the demo window accessible in my application and don't have any
trouble typing into that Hello World box or using any of the other widgets.
I'm using a version of ImGui with the modifications from PR#30 on Kubuntu
22.04.

I'll build the stock vsgImGui and see if there is a problem.

One idea - are you letting vsgImgui pull its own version of ImGui via
CMake? I noticed that they have changed keyboard handling in the latest
version.

Roland


[image: image.png]

**Reply by kannode** [reply]
You have a point. I believe it may be picking up the imgui from my vcpkg toolchain. Let me figure this out some more before going on a hunt for a nonexistent problem.

**Comment by rolandhill** [comment]
My ImGui has suddenly become 1.89.2. The submodule is pulling master - I
thought it was a specific version. I was using version 1.88 before and it
was working fine.

**Comment by rolandhill** [comment]
If I build using ImGui 1.89.2 (master) then I get an assert failure too.

**Comment by robertosfield** [comment]
Curious findings. vsgImGui pulls in a specific commit for ImGui 1.88 from the ImGui repo.  Are you guys modifying your local checkout of vsgImGui to change this, or is vsgImGui somehow pulling in another version of ImGui even though it has a local checkout and build of it:?

On Thursday afternoon I'll have wrapped up my client work for the week and can look into these issue further then.

**Comment by rolandhill** [comment]
I don't recall modifying it, but must have. I just cloned a fresh
vsgImGui and it cloned ImGui 1.88.

**Reply by kannode** [reply]
Some more info.
- Yes the submodule is locked to 9aae45eb (1.88) as we expect.
- On windows, no vcpkg, clean checkout of vsg, vsgimgui and vsgexamples still results in an assertion failure.
- I do not recall modifying locally vsgui locally.
- I feel the vsgImGui/CMakeLists.txt which copies imgui.h using execute_process into vsgimgui/imgui.h and then to install locations might be contributing to the mismatch in headers.
- By the time install happens there are three copies of imgui.h, one in the submodule source location, then in include/vsgImgui/imgui.h during development and then in CMAKE_INSTALL_PREFIX/vsgimgui/imgui.h. 
- Will look at it again after Friday.

**Reply by robertosfield** [reply]
Thanks for the details.  The mystery continues as to why different platforms are behaving differently.

The management of the imgui.h header shouldn't affect things, as long as there's no bug in the update process.

The vsgimgui_example should only ever see the installed include/vsgImGui/imgui.h and never references the vsgImGui source code tree.  If one manually adds the headers to the vsgImGui source code when building an example/3rd party application then things could get confused, but this would be rather perverse given CMake's find_pacakage() should handle this all automatically for you.
